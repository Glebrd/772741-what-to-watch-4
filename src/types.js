import PropTypes from "prop-types";

export const movieType = PropTypes.shape({
  title: PropTypes.string,
  picture: PropTypes.string,
  genre: PropTypes.string,
  date: PropTypes.number,
  poster: PropTypes.string,
  background: PropTypes.string,
  rating: PropTypes.number,
  scores: PropTypes.number,
  director: PropTypes.string,
  starring: PropTypes.arrayOf(PropTypes.string),
  description: PropTypes.string,
  videoPreview: PropTypes.string,
  runTime: PropTypes.number,
  comments: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    user: PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
    }),
    rating: PropTypes.number,
    comment: PropTypes.string,
    date: PropTypes.string,
  }))
});

export const userType = PropTypes.shape({
  id: PropTypes.number,
  email: PropTypes.string,
  name: PropTypes.string,
  avatarURL: PropTypes.string,
});

