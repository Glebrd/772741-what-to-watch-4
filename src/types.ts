
export interface MovieType {
  id: number;
  title: string;
  picture: string;
  genre: string;
  date: number;
  poster: string;
  background: string;
  rating: number;
  scores: number;
  director: string;
  starring: string[];
  description: string;
  videoPreview: string;
  runTime: number;
  backgroundColor: string;
  isFavorite: boolean;
  videoLink: string;
}

export interface UserType {
  id: number;
  email: string;
  name: string;
  avatarURL: string;
}

export interface CommentType {
  id: number;
  user: {
    id: number,
    name: string,
  };
  rating: number;
  comment: string;
  date: string;
}

