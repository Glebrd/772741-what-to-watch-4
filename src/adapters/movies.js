export const adaptMovie = (movie) => {
  return {
    id: movie.id,
    title: movie.name,
    poster: movie.poster_image,
    picture: movie.preview_image,
    background: movie.background_image,
    backgroundColor: movie.background_color,
    videoLInk: movie.video_link,
    videoPreview: movie.preview_video_link,
    description: movie.description,
    rating: movie.rating,
    scores: movie.scores_count,
    director: movie.director,
    starring: movie.starring,
    runTime: movie.run_time,
    genre: movie.genre,
    date: movie.released,
    isFavorite: movie.is_favorite,
  };
};

export const adaptMovies = (movies) => movies.map((movie) => adaptMovie(movie));
