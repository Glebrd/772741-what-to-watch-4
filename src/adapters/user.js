export const adaptUser = (user) => {
  return {
    id: user.id,
    email: user.email,
    name: user.name,
    avatarURL: user.avatar_url,
  };
};
