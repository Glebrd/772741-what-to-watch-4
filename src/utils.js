import {ISOStringProperty, MILLISECONDS_IN_SECOND, MINUTES_IN_HOUR} from "./const";

const DELETE_COUNT = 1;

const TimeAbbreviation = {
  HOURS: `h`,
  MINUTES: `m`,
};

export const extend = (firstSource, secondSource) => {
  return Object.assign({}, firstSource, secondSource);
};

export const formatTime = (seconds) => {
  return new Date(seconds * MILLISECONDS_IN_SECOND).toISOString()
    .substr(ISOStringProperty.START_OF_HOURS, ISOStringProperty.LENGTH_OF_HH_MM_SS_SUBSTRING);
};

export const formatDuration = (totalMinutes) =>{
  let hours = Math.floor(totalMinutes / MINUTES_IN_HOUR);
  let minutes = totalMinutes % MINUTES_IN_HOUR;
  return `${hours + TimeAbbreviation.HOURS} ${minutes + TimeAbbreviation.MINUTES}`;
};

export const replaceMovie = (movie, movies) => {
  const index = movies.findIndex((oldMovie) => oldMovie.id === movie.id);
  movies.splice(index, DELETE_COUNT, movie);
  return movies;
};

export const replacePromo = (updatedMovie, promoMovie) => {
  return updatedMovie.id === promoMovie.id ? updatedMovie : promoMovie;
};


