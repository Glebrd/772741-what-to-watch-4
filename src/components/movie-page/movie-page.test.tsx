import * as React from "react";
import * as renderer from "react-test-renderer";
import MoviePage from "./movie-page";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {adaptMovie, adaptMovies} from "../../adapters/movies";
import {Router} from "react-router-dom";
import history from "../../history";
import {noOperation} from "../../utils";
import {CommentType, MovieType} from "../../types";
import thunk from 'redux-thunk'

const mockStore = configureStore([])

const movies = [
  {
    id: 1,
    title: `1+1`,
    picture: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Comedy`,
    date: 2010,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    background: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    rating: 9.5,
    scores: 898,
    director: `Gardener Grigoriev`,
    starring: [`Maible Rosling`, `Tarrah Sillett`, `Christoper Dunford`, `Helga Stammer`],
    description: `After he becomes a quadriplegic from a paragliding accident, an aristocrat hires a young man from the projects to be his caregiver.`,
    videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    backgroundColor: `#D8E3E5`,
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    runTime: 135,
    isFavorite: true,
  },
  {
    id: 35,
    title: `Inception`,
    picture: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    genre: `Comedy`,
    date: 2010,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    background: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    rating: 9.1,
    scores: 456,
    director: `Evan Ruit`,
    starring: [`Levi Reisen`, `Waldon Verner`, `Ardith Bogace`],
    description: `A thief who steals corporate secrets through the use of dream-sharing technology is given the inverse task of planting an idea into the mind of a C.E.O.`,
    videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    backgroundColor: `#D8E3E5`,
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    runTime: 135,
    isFavorite: true,
  },
];

const movie: MovieType = {
  id: 478,
  title: `Pulp Fiction`,
  picture: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  genre: `Crime`,
  date: 1994,
  poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  background: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
  rating: 7.9,
  scores: 845,
  director: `Kathi Gasticke`,
  starring: [`Adolphus McIndoe`, `Carly Spilsted`, `Rosalind Goodger`, `Teodorico Levey`],
  description: `The lives of two mob hitmen, a boxer, a gangster and his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
  videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  backgroundColor: `#D8E3E5`,
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
  runTime: 135,
  isFavorite: true,
};

const store = mockStore({
  data: {movies: adaptMovies(movies), promoMovie: adaptMovie(movie)},
  application: {currentGenre: `All genres`},
  user: {user: {avatarURL: `img/1.png`}},
});

const user = {
  id: 1,
    email: 'sdfsd@gmail.com',
    name: 'sdfsd',
    avatarURL: '/wtw/static/avatar/9.jpg'
};

const comment:CommentType =
  {
    "id": 1,
    "user": {
      "id": 5,
      "name": `Kate Mur`
    },
    "rating": 8.9,
    "comment": `Two detectives, a rookie and a veteran, hunt a serial killer who uses`,
    "date": `2019-05-08T14:13:56.569Z`
  };

const comments = [comment];

jest.mock('../movie-card-tabs/movie-card-tabs', () => () => {
  const ComponentToMock = () => <div />;
  return `ComponentToMock`;
});

// const MoviePage = require('../movie-card-tabs/movie-card-tabs').default;

// window['onReviewsTabClick'] = jest.fn();

// jest.mock(`../movie-card-tabs/movie-card-tabs.tsx`, () => () => `MovieCardTabs`)

jest.mock('../movie-card-tabs/movie-card-tabs.tsx', () => {
  return {
    'default': 'ChildComponent'
  }
});;

// jest.mock("../movie-card-tabs/movie-card-tabs", function() => () => {
//   return {
//     onReviewsTabClick: jest.fn().mockImplementationOnce(() => Promise.resolve(``)),
//     }
//   });

// jest.mock('../movie-card-tabs/movie-card-tabs', () => ({
//   onReviewsTabClick: jest.fn().mockImplementationOnce(() => Promise.resolve(``)),
// }));

// const onReviewsTabClick = jest.fn().mockImplementationOnce(() => Promise.resolve(``));
// const api = {};
it(`Movie page renders`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <Provider store={store}>
            <MoviePage
              currentMovie={movie}
              user={user}
              sameGenreMovies={movies}
              onPlayClick={noOperation}
              authorizationStatus={`1`}
              onButtonListClick={null}
              match={{params:{id:1}}}
            />
          </Provider>
        </Router>).toJSON();

  expect(tree).toMatchSnapshot();
});
