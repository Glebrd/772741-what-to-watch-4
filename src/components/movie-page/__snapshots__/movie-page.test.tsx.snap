// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie page renders 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <a
            className="user-block__link"
            href="/login"
            onClick={[Function]}
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
             
             
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            />
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            width="218"
          />
        </div>
        <ChildComponent
          movie={
            Object {
              "background": undefined,
              "backgroundColor": undefined,
              "date": undefined,
              "description": "After he becomes a quadriplegic from a paragliding accident, an aristocrat hires a young man from the projects to be his caregiver.",
              "director": "Gardener Grigoriev",
              "genre": "Comedy",
              "id": 1,
              "isFavorite": undefined,
              "picture": undefined,
              "poster": undefined,
              "rating": 9.5,
              "runTime": undefined,
              "scores": undefined,
              "starring": Array [
                "Maible Rosling",
                "Tarrah Sillett",
                "Christoper Dunford",
                "Helga Stammer",
              ],
              "title": undefined,
              "videoLink": undefined,
              "videoPreview": undefined,
            }
          }
        />
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
